### DISCUSSION TITLE
How to offset a Mesh?

### DISCUSSION THREAD

**Discussion by M8kmyday** [mfem/mfem#4638] [Q&A]
For debugging, I would like to offset a 3D Mesh by moving it in 3D space by a distance given by delx, dely, and delz.  Are these the only two steps that are required or is something else needed?
1. loop through the vertices and apply the offset like this:
```
void offsetMesh (Mesh *mesh, real_t delx, real_t dely, real_t delz)
{
   int i=0;
   while (i < mesh->GetNV()) {
      real_t *vertex=mesh->GetVertex(i);
      vertex[0]+=delx;
      vertex[1]+=dely;
      vertex[2]+=delz;
      i++;
   }
}
```
2. call FinalizeMesh

**Comment by tzanio** [comment]
If the mesh is high order, the vertex coordinates don't matter (the `nodes` grid function does) and the above will not be enough.

The best way in general is to use `Mesh::Transform` instead. This will work in both cases and will allow for more general geometry transformations.

If you have to do it once, a convenient way may be to use [Mesh Explorer](https://github.com/mfem/mfem/blob/master/miniapps/meshing/mesh-explorer.cpp) by first modifying its internal [`transformation()`](https://github.com/mfem/mfem/blob/master/miniapps/meshing/mesh-explorer.cpp#L44) function.

**Reply by pazner** [reply]
As @tzanio mentioned, the best way to do this is with `Mesh::Transform`. In your example, you can do the following:

```c++
VectorFunctionCoefficient transf_coeff(
   3, [delx,dely,delz](const Vector &x_in, Vector &x_out)
{
   x_out[0] = x_in[0] + delx;
   x_out[1] = x_in[1] + dely;
   x_out[2] = x_in[2] + delz;
});
mesh.Transform(transf_coeff);
```
